datafusion/src/physical_plan/hash_join.rs

```diff
@@ -1666,6 +1689,41 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn join_semi() -> Result<()> {
+        let left = build_table(
+            ("a1", &vec![1, 2, 3]),
+            ("b1", &vec![4, 5, 7]), // 7 does not exist on the right
+            ("c1", &vec![7, 8, 9]),
+        );
+        let right = build_table(
```

Comment:
It may be worth testing with record batches that have duplicates on the `left` as well (to ensure the duplicates on the left are produced) ?